### DISCUSSION TITLE
Add support for a "next" package.json exports condition

### DISCUSSION THREAD

**Discussion by devongovett** [vercel/next.js#54528] [Ideas]
### Goals

1. Allow libraries to provide Next.js specific versions of modules.
2. Improve DX so that libraries can use Next.js router features automatically, without requiring users to install an adapter manually.

### Background

In [React Aria](https://react-spectrum.adobe.com/react-aria/index.html) we would like to provide deep integration in several of our components with routers in popular frameworks like Next.js. For example, imagine a Tabs component that automatically knew which tab is selected based on the url, and could perform a client side navigation when a user selected a new tab. In order to do this today, users must write manual glue code to call `useRouter` / `usePathname` and provide the current pathname and a function to perform a navigation when the selected tab changes. This is ok and won't be going away but we think we can improve the DX by automatically integrating with framework routers. The problem is we cannot assume React Aria is running in Next.js, so we cannot call `useRouter` ourselves today.

### Proposal

If Next.js supported a "next" package exports condition, we could provide something like this in our packages:

```json
{
  "name": "@react-aria/tabs",
  "exports": {
    "next": "./next-adapter.js",
    "default": "./default.js"
  }
}
```

In `next-adapter.js`, we could then import `useRouter` from `next/navigation` and automatically provide a Next.js specific version of tabs that could automatically determine the current tab from the URL and perform navigations automatically when the selection changes. For non-Next.js environments, the default version would be used that requires devs to manually pass in that info. We could also provide versions for other frameworks like Remix if they offered a similar exports condition.

I think this would be pretty easy to implement on the Next.js side, just a webpack option for the resolver. I'm happy to send a PR if this is a feature that you'd like to support.

**Comment by GeoffreyBooth** [comment]
There’s a registry for these keys, you can submit a PR to get a new one added: https://runtime-keys.proposal.wintercg.org/
